# **Bài 1. Tạo Tensor cơ bản**

Tạo một Tensor số nguyên kích thước 2 × 3 và hiển thị kết quả.

In [14]:
import torch

a = torch.tensor(
    [[1,2,3],
    [4,5,6]],
    dtype=torch.int32
)
print(a)
print("\n")
print("kích thước:",a.shape)
print("Số chiều: ",a.ndim)
print("Kiểu dữ liệu: ", a.dtype)

tensor([[1, 2, 3],
        [4, 5, 6]], dtype=torch.int32)


kích thước: torch.Size([2, 3])
Số chiều:  2
Kiểu dữ liệu:  torch.int32


# **Bài 2. Tạo các Tensor thường dùng**

Thực hành tạo Tensor ngẫu nhiên, Tensor toàn số 0, toàn số 1 và ma trận đơn vị.

In [15]:
import torch
a = torch.rand(3, 4)
b = torch.randn(3, 4)
c = torch.zeros(2, 3)
d = torch.ones(2, 3)
e = torch.eye(4)
f = torch.full((5, 5), 10)

print("rand:\n", a)
print("randn:\n", b)
print("zeros:\n", c)
print("ones:\n", d)
print("eye:\n", e)
print("Tensor 5x5 toàn 10:\n", f)

rand:
 tensor([[0.8445, 0.3046, 0.2953, 0.4057],
        [0.5210, 0.2554, 0.8984, 0.7127],
        [0.6165, 0.2855, 0.8551, 0.2715]])
randn:
 tensor([[-0.3942, -0.3209, -1.5947,  1.3983],
        [ 0.5624, -0.2520,  2.7865,  1.9435],
        [ 0.5464,  0.4055, -1.1518,  1.4103]])
zeros:
 tensor([[0., 0., 0.],
        [0., 0., 0.]])
ones:
 tensor([[1., 1., 1.],
        [1., 1., 1.]])
eye:
 tensor([[1., 0., 0., 0.],
        [0., 1., 0., 0.],
        [0., 0., 1., 0.],
        [0., 0., 0., 1.]])
Tensor 5x5 toàn 10:
 tensor([[10, 10, 10, 10, 10],
        [10, 10, 10, 10, 10],
        [10, 10, 10, 10, 10],
        [10, 10, 10, 10, 10],
        [10, 10, 10, 10, 10]])


# Yêu cầu: Nhận xét sự khác nhau giữa torch.rand() và torch.randn()?
- torch.rand(): tạo tensor với các giá trị ngẫu nhien tuân theo phân phối đêu trong khoảng [0,1). với các giá trị luôn dương
- torch.randn(): Tạo tensor với các giá trị ngẫu nhiên tuân theo phân phối chuẩn có giá tị trung bình bằng 0 và phương sai bằng 1. các giá trị có thể bao gồm cả số âm và dương. 

# **Bài 3. Kiểm tra thông tin Tensor**

Sử dụng shape, size(), ndim và dtype để kiểm tra các thuộc tính của Tensor.
Yêu cầu: giải thích ý nghĩa của torch.size([2,3,4])

In [16]:
import torch
a = torch.zeros(2, 3, 4)
print("Tensor:")
print(a)
print("Shape:", a.shape)
print("Size:", a.size())
print("Number of dimensions:", a.ndim)
print("Data type:", a.dtype)

Tensor:
tensor([[[0., 0., 0., 0.],
         [0., 0., 0., 0.],
         [0., 0., 0., 0.]],

        [[0., 0., 0., 0.],
         [0., 0., 0., 0.],
         [0., 0., 0., 0.]]])
Shape: torch.Size([2, 3, 4])
Size: torch.Size([2, 3, 4])
Number of dimensions: 3
Data type: torch.float32


# Giải thích: Torch.size([2,3,4]):
- đây là 1 cấu trúc tensor 3 chiều(3D)
- 2: tensor này chứa 2 khối (hoặc 2 ma trận con)
- 3: mỗi khối/ ma trận con có 3 hàng
- 4: mỗi khối/ma trận con có 4 cột
- tổng số lượng phần tử chứa có tensor này là 2*3*4 = 24 phần tử

# **Bài 4. Slicing và Indexing**

Truy xuất một phần dữ liệu từ Tensor nhiều chiều.

In [17]:
import torch

# 1. Code mẫu từ đề bài
a = torch.randn(3, 4, 5)
print("Original tensor shape:", a.shape)
print("Original tensor:\n", a)

print("\nTensor a[1]:\n", a[1])
print("Shape a[1]:", a[1].shape)

print("\nTensor a[1:, 2:4]:\n", a[1:, 2:4])
print("Shape a[1:, 2:4]:", a[1:, 2:4].shape)

# 2. Thực hiện yêu cầu: Lấy hai phần tử cuối cùng trên chiều thứ ba
# Dùng slicing [..., -2:] hoặc [:, :, -2:]
last_two_elements = a[:, :, -2:]
print("\nHai phan tu cuoi tren chieu thu ba (shape):", last_two_elements.shape)
print("Gia tri:\n", last_two_elements)

Original tensor shape: torch.Size([3, 4, 5])
Original tensor:
 tensor([[[ 1.5388, -0.3714,  0.5382, -1.4686,  0.6104],
         [-1.3148, -0.4223, -0.2051, -0.2019,  0.4691],
         [-0.9862,  1.2810, -1.2203, -0.2896,  0.7848],
         [ 0.5066, -1.9587,  0.5945,  0.7297,  0.8240]],

        [[-0.0988,  1.1734, -0.2459,  2.3237, -0.5166],
         [ 1.9194,  1.4951,  0.1714,  0.7577,  1.0668],
         [-0.3776, -0.8923,  0.7547,  0.0601,  1.0776],
         [-1.4900,  0.3343,  2.2337,  0.1949, -1.0922]],

        [[ 0.9555,  0.2833, -0.1441,  1.2576, -0.6149],
         [ 0.6403, -0.7464, -0.8459,  0.5592, -1.6481],
         [ 0.0238, -1.7522,  2.5057, -1.9596, -1.6757],
         [ 1.9151, -0.8056,  0.4160,  0.9786,  0.7589]]])

Tensor a[1]:
 tensor([[-0.0988,  1.1734, -0.2459,  2.3237, -0.5166],
        [ 1.9194,  1.4951,  0.1714,  0.7577,  1.0668],
        [-0.3776, -0.8923,  0.7547,  0.0601,  1.0776],
        [-1.4900,  0.3343,  2.2337,  0.1949, -1.0922]])
Shape a[1]: torch.Size(

Yêu cầu:
• Cho biết shape của a[1] và a[1:, 2:4].
• Thử lấy hai phần tử cuối cùng trên chiều thứ ba.

Hình dạng (shape) của các lát cắt (slicing):   
- a[1]: Có shape là torch.Size([4, 5]). Khi cố định chỉ số ở chiều thứ nhất (index = 1), chiều này bị tiêu biến, để lại ma trận 2 chiều gồm 4 hàng và 5 cột.
- a[1:, 2:4]: Có shape là torch.Size([2, 2, 5]). Ta cắt lấy các khối từ chỉ số 1 trở đi (chiều 0 còn 2 khối), lấy các hàng từ 2 đến trước 4 (chiều 1 còn 2 hàng), và giữ nguyên toàn bộ 5 cột ở chiều cuối cùng.

Lấy hai phần tử cuối cùng trên chiều thứ ba:   Cú pháp: a[:, :, -2:] (hoặc a[..., -2:]).
- Ý nghĩa: Giữ nguyên tất cả các khối (chiều 0) và tất cả các hàng (chiều 1), chỉ lấy 2 phần tử cuối cùng ở chiều cột (chiều 2) thông qua chỉ số âm -2:. Kết quả trả về một Tensor có kích thước torch.Size([3, 4, 2]).   

# **Bài 5. Reshape và Flatten**

Thay đổi hình dạng Tensor mà không làm thay đổi số phần tử.

In [18]:
import torch

a = torch.randn(3,4)

print("Original shape:", a.shape)
print(a)

print("\nFlatten:", a.ravel())
print("Flatten shape:", a.ravel().shape)

reshaped_322 = a.reshape(3, 2, 2)
print("\nReshape to 3 x 2 x 2:\n", reshaped_322)
print("Reshaped (3, 2, 2) shape:", reshaped_322.shape)

# Yêu cầu: Kiểm tra số phần tử và reshape thành 2 x 6
print("\n--- Kiem tra so phan tu ---")
print("So phan tu ban dau (a.numel()):", a.numel())
print("So phan tu sau reshape 3x2x2:", reshaped_322.numel())

reshaped_26 = a.reshape(2, 6)
print("\nReshape to 2 x 6:\n", reshaped_26)
print("Reshaped (2, 6) shape:", reshaped_26.shape)
print("So phan tu sau reshape 2x6:", reshaped_26.numel())

Original shape: torch.Size([3, 4])
tensor([[ 0.2429,  0.5749,  2.1305, -1.3628],
        [-0.1706, -0.0432, -1.6814, -0.5928],
        [-1.1445, -0.1835,  1.0862, -0.3283]])

Flatten: tensor([ 0.2429,  0.5749,  2.1305, -1.3628, -0.1706, -0.0432, -1.6814, -0.5928,
        -1.1445, -0.1835,  1.0862, -0.3283])
Flatten shape: torch.Size([12])

Reshape to 3 x 2 x 2:
 tensor([[[ 0.2429,  0.5749],
         [ 2.1305, -1.3628]],

        [[-0.1706, -0.0432],
         [-1.6814, -0.5928]],

        [[-1.1445, -0.1835],
         [ 1.0862, -0.3283]]])
Reshaped (3, 2, 2) shape: torch.Size([3, 2, 2])

--- Kiem tra so phan tu ---
So phan tu ban dau (a.numel()): 12
So phan tu sau reshape 3x2x2: 12

Reshape to 2 x 6:
 tensor([[ 0.2429,  0.5749,  2.1305, -1.3628, -0.1706, -0.0432],
        [-1.6814, -0.5928, -1.1445, -0.1835,  1.0862, -0.3283]])
Reshaped (2, 6) shape: torch.Size([2, 6])
So phan tu sau reshape 2x6: 12


# Trả lời yêu cầu:

- Số phần tử trước reshape: 12
- Số phần tử sau reshape: 12



# **Bài 6. Phép toán và nhân ma trận**
Thực hiện các phép toán phần tử và phép nhân ma trận bằng toán tử @.

In [19]:
import torch

a = torch.randn(2, 3)
b = torch.randn(2, 3)

print("a:\n", a)
print("b:\n", b)
print("\na + b:\n", a + b)
print("\na^2:\n", a ** 2)
print("\na / b:\n", a / b)

# Nhan tung phan tu vs Nhan ma tran
elementwise_prod = a * b
matrix_mul = a @ b.T

print("\nElement-wise multiplication (a * b):\n", elementwise_prod)
print("\nMatrix multiplication (a @ b.T):\n", matrix_mul)
print("Shape cua a @ b.T:", matrix_mul.shape)

a:
 tensor([[-1.4445,  0.3277, -0.7330],
        [-0.4560, -0.2524, -0.5009]])
b:
 tensor([[-0.3792, -0.6047, -0.5455],
        [-0.1200,  0.7115,  0.0999]])

a + b:
 tensor([[-1.8237, -0.2771, -1.2785],
        [-0.5760,  0.4592, -0.4011]])

a^2:
 tensor([[2.0867, 0.1074, 0.5373],
        [0.2080, 0.0637, 0.2509]])

a / b:
 tensor([[ 3.8098, -0.5418,  1.3437],
        [ 3.8003, -0.3547, -5.0164]])

Element-wise multiplication (a * b):
 tensor([[ 0.5477, -0.1981,  0.3999],
        [ 0.0547, -0.1796, -0.0500]])

Matrix multiplication (a @ b.T):
 tensor([[ 0.7494,  0.3333],
        [ 0.5988, -0.1749]])
Shape cua a @ b.T: torch.Size([2, 2])


# Trả lời yêu cầu
Phân biệt a * b và a @ b.T
- a * b : phép nhân từng phần tử tương ứng giữa 2 tensor cùng kích thước (2x3), kết quả ra shape (2,3)
- a @ b.T : Phép nhân ma trận đại số tuyến tính. Ma trận a kích thước (2 x 3) nhân với chuyển vị của b kích thước (3x2)

Shape của kết quả a @ b.T là torch.Size([2, 2])

# **Bài 7. Các hàm thống kê trên Tensor**

Tính trung bình, độ lệch chuẩn và tổng tích lũy theo từng cột.

In [20]:
import torch

a = torch.randn(3,4)
print("Tensor a:\n", a)

print("\n--- Theo cot (dim = 0)---")
print("Mean (dim=0):", torch.mean(a, dim=0))
print("Std (dim=0):", torch.std(a, dim=0))
print("Cumsum (dim=0):\n", torch.cumsum(a, dim=0))

print("\n--- Theo hang (dim=1) ---")
print("Mean (dim=1):", torch.mean(a, dim=1))
print("Std (dim=1):", torch.std(a, dim=1))
print("Cumsum (dim=1):\n", torch.cumsum(a, dim=1))

Tensor a:
 tensor([[ 1.4902, -0.1589,  0.3596, -0.9312],
        [-0.4968,  2.1109,  0.0314,  0.2560],
        [ 0.8370,  1.7660,  0.8716,  0.4096]])

--- Theo cot (dim = 0)---
Mean (dim=0): tensor([ 0.6101,  1.2393,  0.4209, -0.0885])
Std (dim=0): tensor([1.0127, 1.2231, 0.4235, 0.7338])
Cumsum (dim=0):
 tensor([[ 1.4902, -0.1589,  0.3596, -0.9312],
        [ 0.9934,  1.9520,  0.3910, -0.6752],
        [ 1.8303,  3.7180,  1.2626, -0.2656]])

--- Theo hang (dim=1) ---
Mean (dim=1): tensor([0.1899, 0.4754, 0.9710])
Std (dim=1): tensor([1.0162, 1.1351, 0.5701])
Cumsum (dim=1):
 tensor([[ 1.4902,  1.3313,  1.6909,  0.7596],
        [-0.4968,  1.6141,  1.6455,  1.9015],
        [ 0.8370,  2.6029,  3.4745,  3.8841]])


Yêu cầu: Thay dim=0 bằng dim=1 và nhận xét sự khác nhau.

Khác biệt giữa dim=0 và dim=1:
- dim=0: Thực hiện phép tính dọc theo các hàng (tức tính trên từng cột), triệt tiêu chiều hàng để trả về kết quả kích thước 4 phần tử tương ứng với 4 cột.
- dim=1: Thực hiện phép tính ngang qua các cột (tức tính trên từng hàng), triệt tiêu chiều cột để trả về kết quả kích thước 3 phần tử tương ứng với 3 hàng.   

# **Bài 8. Autograd cơ bản**

Tạo biến có requires_grad=True và dùng backward() để tính đạo hàm tự động.

In [21]:
import torch

# 1. Ham y = x * x
x = torch.tensor(3.6, requires_grad=True)
y = x * x
y.backward()
print("x =", x)
print("y =", y)
print("x.grad =", x.grad)

# 2. Yeu cau: Doi thanh y = 3*x**2 + 2*x + 1
x2 = torch.tensor(3.6, requires_grad=True)
y2 = 3 * x2**2 + 2 * x2 + 1
y2.backward()
print("\n--- Ham y = 3*x^2 + 2*x + 1 ---")
print("x2.grad =", x2.grad)

x = tensor(3.6000, requires_grad=True)
y = tensor(12.9600, grad_fn=<MulBackward0>)
x.grad = tensor(7.2000)

--- Ham y = 3*x^2 + 2*x + 1 ---
x2.grad = tensor(23.6000)


# Trả lời yêu cầu:

1. Tính thủ công đạo hàm của y = x2 tại x = 3.6 và so sánh với x.grad.

   Đạo hàm thủ công hàm y = x2 tại x = 3.6

       y'(x) = 2x => y'(3.6) = 2 x 3.6 = 7.2

  -> Giá trị trùng khớp hoàn toàn với x.grad thu được từ pytorch(7.2000)

2. Đổi hàm thành y = 3*x**2 + 2*x + 1 và kiểm tra gradient.

   Đạo hàm thủ công hàm y = 3*x**2 + 2*x + 1 tại x = 3.6

       y'(x) = 6x + 2 => y'(3.6) = 6 x 3.6 + 2 = 23.6
   
   Kết quả autograd x2.grad tính ra chính xác 23.6000

# **Bài 9. Gradient Descent tìm hệ số đa thức**

Sinh dữ liệu từ đa thức y = x2 + 2x + 3, sau đó dùng NAdam và autograd để tìm lại ba hệ số.

In [22]:
import numpy as np
import torch

polynomial = np.poly1d([1, 2, 3])
N = 20
X = np.random.randn(N, 1) * 5
Y = polynomial(X)

XX = np.hstack([X * X, X, np.ones_like(X)])
w = torch.randn(3, 1, requires_grad=True)
x = torch.tensor(XX, dtype=torch.float32)
y = torch.tensor(Y, dtype=torch.float32)

optimizer = torch.optim.NAdam([w], lr=0.01)

print("Initial coefficients:\n", w.data)

for epoch in range(1000):
    optimizer.zero_grad()
    y_pred = x @ w
    mse = torch.mean(torch.square(y - y_pred))
    mse.backward()
    optimizer.step()

print("\nLearned coefficients:\n", w.data)

Initial coefficients:
 tensor([[ 0.4039],
        [-0.6568],
        [-2.0289]])

Learned coefficients:
 tensor([[1.0445],
        [2.0502],
        [1.6811]])


Yêu cầu:

1. So sánh ba hệ số học được với [1, 2, 3] : Sau 1000 epoch huấn luyện, các hệ số trong ma trận $w$ hội tụ rất sát với các giá trị thực tế $A \approx 1$, $B \approx 2$, $C \approx 3$.   

2. Giải thích vai trò của optimizer.zero_grad(), mse.backward() và optimizer.step()?
- optomizer.zero_grad(): xóa sạch các gradient tích lũy từ các vòng lặp trước đó về 0 trước khi tính đạo hàm mới
- mse.backward(): Kích hoạt lan truyền ngược để tính toán gradient của hàm mất mát mse đối với trọng số $w$($\frac{\partial \text{mse}}{\partial w}$).
- optimizer.step(): Cập nhật giá trị các trọng số $w$ theo hướng giảm dần của gradient dựa vào thuật toán tối ưu NAdam

# **Bài 10. Giải hệ phương trình bằng Autograd**

Tối thiểu hóa tổng sai số bình phương để tìm A, B, C, D thỏa bốn phương trình.

In [23]:
import random
import torch

A = torch.tensor(random.random(), requires_grad=True)
B = torch.tensor(random.random(), requires_grad=True)
C = torch.tensor(random.random(), requires_grad=True)
D = torch.tensor(random.random(), requires_grad=True)

EPOCHS = 2000
optimizer = torch.optim.NAdam([A, B, C, D], lr=0.01)

for _ in range(EPOCHS):
    y1 = A + B - 9
    y2 = C - D - 1
    y3 = A + C - 8
    y4 = B - D - 2
    sqerr = y1*y1 + y2*y2 + y3*y3 + y4*y4
    
    optimizer.zero_grad()
    sqerr.backward()
    optimizer.step()

print(f"A = {A.item():.4f}")
print(f"B = {B.item():.4f}")
print(f"C = {C.item():.4f}")
print(f"D = {D.item():.4f}")

# Kiem tra tong sai so cuoi cung
print("Final loss (sqerr):", sqerr.item())

A = 4.3553
B = 4.6446
C = 3.6446
D = 2.6446
Final loss (sqerr): 3.2783304959593806e-09


Trả lời yêu cầu:

• Kiểm tra nghiệm thu được với bốn phương trình: A+B=9, C-D=1, A+C=8, B-D=2.
- $A + B \approx 9$

- $C - D \approx 1$

- $A + C \approx 8$

- $B - D \approx 2$


-> Hàm mất mát sqerr tiệm cận về 0, chứng minh nghiệm tìm được hoàn toàn thỏa mãn.

• Nhận xét vì sao bài toán có thể có nhiều nghiệm:

Coi hệ 4 phương trình tuyến tính 4 ẩn:

$$\begin{cases} A + B = 9 \\ C - D = 1 \\ A + C = 8 \\ B - D = 2 \end{cases}$$

Lấy phương trình (1) trừ phương trình (4):
$(A + B) - (B - D) = 9 - 2 \implies A + D = 7$.

Cộng phương trình (2) và (3): $(C - D) + (A + C) = 1 + 8 \implies A + 2C - D = 9$.

Thực chất phương trình (4) phụ thuộc tuyến tính vào 3 phương trình còn lại (hệ có hạng ma trận hệ số bằng 3 nhưng có 4 ẩn). Do đó hệ có bậc tự do bằng 1, dẫn đến việc có vô số nghiệm phụ thuộc vào giá trị khởi tạo ngẫu nhiên ban đầu.


# **3. Bài tập áp dụng**
# Bài 11. Viết chương trình PyTorch để thực hiện các yêu cầu sau:

In [24]:
import numpy as np
import torch

# Yeu cau 1: Tao Tensor X co 24 phan tu va reshape
print("=== YEU CAU 1 ===")
X_tensor = torch.arange(24)
print("Reshape 4x6:\n", X_tensor.reshape(4, 6))
print("Reshape 2x12:\n", X_tensor.reshape(2, 12))
print("Reshape 2x3x4:\n", X_tensor.reshape(2, 3, 4))

# Yeu cau 2: Thao tac tren Tensor A, B kich thuoc 3x4
print("\n=== YEU CAU 2 ===")
A = torch.randn(3, 4)
B = torch.randn(3, 4)
print("A + B:\n", A + B)
print("A * B:\n", A * B)
print("A / B:\n", A / B)
print("Trung binh A theo cot (dim=0):", torch.mean(A, dim=0))
print("Do lech chuan A theo cot (dim=0):", torch.std(A, dim=0))

# Yeu cau 3: Tinh dy/dx cua y = 2x^2 + 5x + 3 tai x = 2
print("\n=== YEU CAU 3 ===")
x_val = torch.tensor(2.0, requires_grad=True)
y_func = 2 * (x_val ** 2) + 5 * x_val + 3
y_func.backward()
print("Autograd dy/dx tai x = 2:", x_val.grad.item())
# Tinh thu cong: dy/dx = 4x + 5 => 4*2 + 5 = 13
print("Tinh thu cong dy/dx: 4*(2) + 5 = 13")

# Yeu cau 4: Huan luyen da thuc y = 2x^2 - 3x + 5
print("\n=== YEU CAU 4 ===")
poly_new = np.poly1d([2, -3, 5])
N = 30
X_poly = np.random.randn(N, 1) * 3
Y_poly = poly_new(X_poly)

XX_poly = np.hstack([X_poly * X_poly, X_poly, np.ones_like(X_poly)])
w_poly = torch.randn(3, 1, requires_grad=True)
x_p = torch.tensor(XX_poly, dtype=torch.float32)
y_p = torch.tensor(Y_poly, dtype=torch.float32)

opt = torch.optim.NAdam([w_poly], lr=0.01)

for epoch in range(1500):
    opt.zero_grad()
    pred = x_p @ w_poly
    loss = torch.mean(torch.square(y_p - pred))
    loss.backward()
    opt.step()

print("He so tim duoc sau khi huan luyen:")
print(f"A = {w_poly[0].item():.4f} (Goc: 2)")
print(f"B = {w_poly[1].item():.4f} (Goc: -3)")
print(f"C = {w_poly[2].item():.4f} (Goc: 5)")

=== YEU CAU 1 ===
Reshape 4x6:
 tensor([[ 0,  1,  2,  3,  4,  5],
        [ 6,  7,  8,  9, 10, 11],
        [12, 13, 14, 15, 16, 17],
        [18, 19, 20, 21, 22, 23]])
Reshape 2x12:
 tensor([[ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11],
        [12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23]])
Reshape 2x3x4:
 tensor([[[ 0,  1,  2,  3],
         [ 4,  5,  6,  7],
         [ 8,  9, 10, 11]],

        [[12, 13, 14, 15],
         [16, 17, 18, 19],
         [20, 21, 22, 23]]])

=== YEU CAU 2 ===
A + B:
 tensor([[-0.1281,  0.0088,  1.3106, -3.1684],
        [-1.9843, -0.4639, -0.0775,  1.2483],
        [ 0.8712, -3.3088, -1.8817, -1.0326]])
A * B:
 tensor([[-2.5787e-01, -6.8125e-02, -3.3002e+00,  2.3219e+00],
        [ 8.9457e-01, -1.5671e-01,  1.2699e-03, -2.9603e-01],
        [-1.0640e+00,  2.7368e+00,  8.8397e-01, -6.4482e-01]])
A / B:
 tensor([[-1.2860, -1.0342, -0.4933,  1.7529],
        [ 1.8655, -3.0448,  0.4354, -7.1237],
        [-2.2736,  0.9841,  0.9276, -3.3556]])
Trung bin

Trả lời yêu cầu:

- Đạo hàm thủ công yêu cầu 3:
$$y'(x) = 4x + 5 \implies y'(2) = 4(2) + 5 = 13$$

Trùng khớp với kết quả autograd trả về $13.0$.   

- Kết quả huấn luyện yêu cầu 4:
Mô hình hội tụ về các hệ số xấp xỉ $[2, -3, 5]$, đúng với đa thức khởi tạo $y = 2x^2 - 3x + 5$.   